### PR TITLE
unix/fs: Fix for potential partial reads/writes

### PR DESCRIPTION
**Pull request by bwijen**
As [mentioned](https://github.com/libuv/libuv/pull/269#discussion_r28131975) by @bnoordhuis.

CC: @saghul

### CODE REVIEW

**Comment by saghul** [comment]
Style nits and commits message compliance aside, can a test for this be written? (also please do a rebase, since neighboring code was modified). Thanks!

**Comment by bwijen** [comment]
Not sure about those tests, I did test of course but with hard-coded partial reads/writes...
Ideas are welcome...

**Comment by saghul** [comment]
> Not sure about those tests, I did test of course but with hard-coded partial reads/writes...
> Ideas are welcome...

Something like this could work I think: https://github.com/libuv/libuv/blob/v1.x/test/test-eintr-handling.c

**Comment by bwijen** [comment]
Those updates are for style, still have to look at the test implementation...

**Comment by bwijen** [comment]
Sorry for the long delay, please take a look at the last commit and let me know if this is what you suggested.

Side note: I ran these tests without my fix, which in fact cause them to fail.

File: src/unix/fs.c [hunk 1]
```diff
@@ -257,7 +257,7 @@ static ssize_t uv__fs_open(uv_fs_t* req) {
 
 static ssize_t uv__fs_read(uv_fs_t* req) {
 #if defined(__linux__)
-  static int no_preadv;
+  static int no_preadv = 0;
```
Comment (saghul) [inline src/unix/fs.c]: No need, it's automagically initialized to 0;

File: src/unix/fs.c [hunk 2]
```diff
@@ -648,7 +630,7 @@ static ssize_t uv__fs_utime(uv_fs_t* req) {
 
 static ssize_t uv__fs_write(uv_fs_t* req) {
 #if defined(__linux__)
-  static int no_pwritev;
+  static int no_pwritev = 0;
```
Comment (saghul) [inline src/unix/fs.c]: ditto

File: test/test-fs.c [hunk 3]
```diff
@@ -2662,3 +2662,104 @@ TEST_IMPL(fs_read_write_null_arguments) {
 
   return 0;
 }
+
+static void thread_exec(int fd, char* data, int size, int interval, int doread) {
+  pid_t pid=getpid();
```
Comment (saghul) [inline test/test-fs.c]: This makes the test Unix only, you need some ifdefs to avoid breaking Windows.

Comment (saghul) [inline test/test-fs.c]: Or better yet, make the test also work on Windows :-)

File: test/test-fs.c [hunk 4]
```diff
@@ -2662,3 +2662,104 @@ TEST_IMPL(fs_read_write_null_arguments) {
...
+  ssize_t result=1;
+
+  while(size>0 && result>0) {
```
Comment (saghul) [inline test/test-fs.c]: General style comment: space between operators.

File: test/test-fs.c (hunk 3 above)
Comment (saghul) [inline test/test-fs.c]: Nah, discard the Windows thing, just add some ifdefs to skip the test on Windows.

**Comment by saghul** [comment]
CI: https://ci.nodejs.org/job/libuv-test-commit/11/

File: src/unix/fs.c (hunk 1 above)
Comment (bwijen) [inline src/unix/fs.c]: It think it's more clear this way..., wouldn't you agree?

Comment (saghul) [inline src/unix/fs.c]: We use this construct all across, I'd go for consistency.

**Comment by saghul** [comment]
- one of the added tests is broken in SmartOS and OSX
- eintr_handling is now broken
- the tests do not compile on Windows

**Comment by bwijen** [comment]
Fixes (all?) style issues. Hopefully disabled the test on windows.
@saghul: Can you please restart a build (I don't seem able to do so.)

There is (probably an intentional) bug in eintr_handling.
The size of the read buffer is to large.
The read stops because of the interrupt, which is exactly what my patch fixes.
(continue reading until all buffers are full, that is.)

I can fix the eintr_handling (with an other interruptable system call: like sleep) but I guess that defeats the purpose of the test.

Comments are welcome...

**Comment by saghul** [comment]
> Can you please restart a build (I don't seem able to do so.)

Only collaborators can trigger builds at the moment. CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/44/

I think this PR should remove the eintr_handling test, since it "fixes" it, right?

**Comment by bwijen** [comment]
@saghul: Yet Another Build Request

EINTR handling is now done inside the loop iso outside, this way it can benefit from the new partial read behaviour.
Removed the EINTR test, since it has been deprecated by the new partial read behaviour.

File: Makefile.am [hunk 5]
```diff
@@ -151,7 +151,6 @@ test_run_tests_SOURCES = test/blackhole-server.c \
                          test/test-default-loop-close.c \
                          test/test-delayed-accept.c \
                          test/test-dlerror.c \
-                         test/test-eintr-handling.c \
```
Comment (saghul) [inline Makefile.am]: you also need to remove it from uv.gyp

File: src/unix/fs.c [hunk 6]
```diff
@@ -828,7 +828,9 @@ static ssize_t uv__fs_buf_iter(uv_fs_t* req, uv__fs_buf_iter_processor process)
     if (req->nbufs > iovmax)
       req->nbufs = iovmax;
 
-    result = process(req);
+    do result = process(req);
```
Comment (saghul) [inline src/unix/fs.c]: please expand this to take 2 lines

File: src/unix/fs.c [hunk 7]
```diff
@@ -844,15 +846,9 @@ static ssize_t uv__fs_buf_iter(uv_fs_t* req, uv__fs_buf_iter_processor process)
...
 
-  if (errno == EINTR && total == -1)
-    return total;
-
   if (bufs != req->bufsml)
     uv__free(bufs);
 
-  req->bufs = NULL;
-  req->nbufs = 0;
-
```
Comment (saghul) [inline src/unix/fs.c]: Don't remove this. We don't want to leave a dangling pointer in the request structure.

File: test/test-fs.c [hunk 8]
```diff
@@ -2684,10 +2684,12 @@ static void thread_exec(int fd, char* data, int size, int interval, int doread)
...
-    if(doread == 0)
-      result = write(fd, data, size<interval?size:interval);
-    else
-      result = read(fd, data, size<interval?size:interval);
+    do {
+      if(doread == 0)
+        result = write(fd, data, size<interval?size:interval);
+      else
+        result = read(fd, data, size<interval?size:interval);
+    }while(result == -1 && errno == EINTR);
```
Comment (saghul) [inline test/test-fs.c]: style: space after }

Comment (txdv) [inline test/test-fs.c]: shouldn't there be a space between if and ( ?

**Comment by bwijen** [comment]
I think I got all of them...

**Comment by saghul** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/48/

**Comment by bwijen** [comment]
So I got all but smartos right?
I don't seem able to run [smartos under VirtualBox](https://wiki.smartos.org/display/DOC/SmartOS+as+a+Sandboxed+VirtualBox+Guest) any ideas?

**Comment by saghul** [comment]
> So I got all but smartos right?

Yep, looks that way!

> I don't seem able to run smartos under VirtualBox any ideas?

TBH I have no idea either :-(

**Comment by santigimeno** [comment]
@bwijen I've been successful to have a working `smartos` virtual machine in the past using KVM and following this guide: https://gist.github.com/AndreasMadsen/3732950. I you have problems just let me know.

**Comment by bwijen** [comment]
So I was able to run smartos under VirtualBox (I used usb image but need live iso)

I ran the fs_partial_write several times, it fails about 1 out of 25.
When it fails, the write (test/test-fs.c:2689) fails with result:=-1 and errno:=0

I tried to write-up a test to reproduce without libuv, but unsuccessful so far.
I can workaround the problem, but I guess that defeats the purpose.

Any suggestions are welcome.

**Comment by bwijen** [comment]
I rebased and added a workaround.
@saghul: could you send these to the buildserver?

**Comment by saghul** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/104/

**Comment by bwijen** [comment]
Both test passed on all builds...
Any comments left?

File: src/unix/fs.c [hunk 9]
```diff
@@ -856,6 +820,19 @@ static int uv__fs_fstat(int fd, uv_stat_t *buf) {
   return ret;
 }
 
+static size_t uv__fs_buf_offset(uv_buf_t* bufs, size_t size) {
+  size_t offset;
+  /* Figure out which bufs are done */
+  for (offset = 0; size > 0 && bufs[offset].len <= size; ++offset) {
+    size -= bufs[offset].len;
+  }
```
Comment (saghul) [inline src/unix/fs.c]: style: drop braces for single statement for

File: src/unix/fs.c [hunk 10]
```diff
@@ -875,7 +852,10 @@ static ssize_t uv__fs_buf_iter(uv_fs_t* req, uv__fs_buf_iter_processor process)
     if (req->nbufs > iovmax)
       req->nbufs = iovmax;
 
-    result = process(req);
+    do
+      result = process(req);
+    while(result < 0 && errno == EINTR);
```
Comment (saghul) [inline src/unix/fs.c]: style: space before (

File: test/test-fs.c [hunk 11]
```diff
@@ -2684,3 +2684,122 @@ TEST_IMPL(fs_read_write_null_arguments) {
...
+}
+
+#else  /* !_WIN32 */
+
+static void thread_exec(int fd, char* data, int size, int interval, int doread) {
+  const pid_t pid = getpid();
+  ssize_t result = 1;
+  int error = 0;
+
+  while(size > 0 && error == 0) {
```
Comment (saghul) [inline test/test-fs.c]: style: space before (

File: test/test-fs.c [hunk 12]
```diff
@@ -2684,3 +2684,122 @@ TEST_IMPL(fs_read_write_null_arguments) {
...
+    do {
+      if(doread)
+        result = write(fd, data, size<interval?size:interval);
```
Comment (saghul) [inline test/test-fs.c]: style: space between operator

File: test/test-fs.c [hunk 13]
```diff
@@ -2684,3 +2684,122 @@ TEST_IMPL(fs_read_write_null_arguments) {
...
+      else
+        result = read(fd, data, size<interval?size:interval);
```
Comment (saghul) [inline test/test-fs.c]: ditto

File: test/test-fs.c [hunk 14]
```diff
@@ -2684,3 +2684,122 @@ TEST_IMPL(fs_read_write_null_arguments) {
...
+
+      error=errno;
```
Comment (saghul) [inline test/test-fs.c]: ditto

File: test/test-fs.c [hunk 15]
```diff
@@ -2684,3 +2684,122 @@ TEST_IMPL(fs_read_write_null_arguments) {
...
+    } while(result == -1 && error == EINTR);
```
Comment (saghul) [inline test/test-fs.c]: style: space before (

File: test/test-fs.c [hunk 16]
```diff
@@ -2684,3 +2684,122 @@ TEST_IMPL(fs_read_write_null_arguments) {
...
+
+    kill(pid, SIGUSR1);
+    if(result > 0) {
```
Comment (saghul) [inline test/test-fs.c]: ditto

File: test/test-fs.c [hunk 17]
```diff
@@ -2684,3 +2684,122 @@ TEST_IMPL(fs_read_write_null_arguments) {
...
+}
+
+struct thread_ctx {
+  int fd;
+  char *data;
+  int size;
+  int interval;
+  int doread;
+};
+static void thread_main(void* arg) {
```
Comment (saghul) [inline test/test-fs.c]: add a newline here

**Comment by saghul** [comment]
Some style nits, otherwise LGTM. Can you also squash the commits into one since you're here? Thanks!

**Comment by bwijen** [comment]
Fixed nits + squashed.

**Comment by saghul** [comment]
LGTM. @libuv/collaborators can anyone have another look at this just in case?

**Comment by saghul** [comment]
Also, question: does Windows also behave like this? If so, we should probably add a note in the documentation, mentioning that starting with 1.10 either all the provided buffers are written or an error occurs.

**Comment by misterdjules** [comment]
@bwijen @saghul Please, feel free to mention @libuv/sunos next time you have an issue with SmartOS :)

**Comment by saghul** [comment]
@misterdjules sure thing! as you may see this PR started a while ago... Mind taking a look, btw?

**Comment by misterdjules** [comment]
@saghul Yes, I'm taking a look right now, I'll keep you posted.

**Comment by misterdjules** [comment]
The only added test that was [mentioned to not pass on SmartOS](https://github.com/libuv/libuv/pull/640#issuecomment-221385266) is `fs_partial_write`, but I've been running it in a loop on SmartOS for the past five minutes and I haven't seen any failure so far.

This might be due to the "workaround" that @bwijen [mentioned earlier](https://github.com/libuv/libuv/pull/640#issuecomment-241274063). Unfortunately I can't see what that workaround is.

@bwijen Do you mind indicating what that workaround is so that I can see if we can come up with a fix that doesn't defeat the purpose of the test on SmartOS?

**Comment by bwijen** [comment]
@misterdjules: Thanks for looking into this!
The 'workaround' can be found here: bwijen/libuv@b0d251ee5730d97c5f5ae67a2df2c5890340f376

Without that last commit the test fails (in my virtual Smart OS environment) about 1 out of 25

**Comment by misterdjules** [comment]
@bwijen I'm running `./test/run-tests fs_partial_write` with the code in this PR and the following patch applied, which is the result of reverting [the workaround you mentioned in your previous comment](https://github.com/bwijen/libuv/commit/b0d251ee5730d97c5f5ae67a2df2c5890340f376):

```diff
commit 15a91094edf68f17ae9dbb718387e716841f1cde
Author: Julien Gilli <jgilli@fastmail.fm>
Date:   Mon Sep 12 21:30:03 2016 +0000

     revert smartos workaround for PR 640

diff --git a/test/test-fs.c b/test/test-fs.c
index 0c5761c..389535b 100644
--- a/test/test-fs.c
+++ b/test/test-fs.c
@@ -2700,25 +2700,21 @@ TEST_IMPL(fs_partial_write) {
 static void thread_exec(int fd, char* data, int size, int interval, int doread) {
   const pid_t pid = getpid();
   ssize_t result = 1;
-  int error = 0;
 
-  while (size > 0 && error == 0) {
+  while (size > 0 && result > 0) {
     do {
       if(doread)
         result = write(fd, data, size < interval ? size : interval);
       else
         result = read(fd, data, size < interval ? size : interval);
 
-      error = errno;
-    } while (result == -1 && error == EINTR);
+    } while (result == -1 && errno == EINTR);
 
     kill(pid, SIGUSR1);
-    if (result > 0) {
-      error = 0;
-      size -= result;
-      data += result;
-    }
+    size -= result;
+    data += result;
   }
+
   ASSERT(size == 0);
   ASSERT(result > 0);
 }
```

However I get only failures, the test fails with the following output:

```
[root@0c933410-7c9b-4bea-96b6-2db35946bc95 ~/libuv]# ./test/run-tests fs_partial_write
not ok 1 - fs_partial_write
# exit code 134
# Output from process `fs_partial_write`:
# Assertion failed in test/test-fs.c on line 2718: size == 0
[root@0c933410-7c9b-4bea-96b6-2db35946bc95 ~/libuv]#
```

Is the patch above the correct revert of your workaround? You mentioned that the test fails sporadically (once out of ~25 runs), so I wouldn't expect the test to fail all the time without your workaround. What am I missing?

**Comment by bwijen** [comment]
@misterdjules:
That patch is indeed the correct revert.
For sure it is possible that it will fail 1 out of 1 on a real SmartOS system (I was running it under VirtualBox)
I guess it would just be easier to pinpoint the problem if it always fails

The problem I saw (only on SmartOS) that result:=-1 and errno:=0 which causes it to break out of the both loops while not all data is sent (hence the assert on size==0)

**Comment by misterdjules** [comment]
@saghul @bwijen

The problem is that on SmartOS, `errno` is not reentrant by default. One has to pass the `-pthreads` option to the compiler in order for `errno` to be reentrant. This is not needed on the CentOS distribution I used to test that behavior on Linux.

The following changes fix the problem (I took that opportunity to change `libuv_la_CFLAGS` from `-pthread` to `-pthreads` as it's consistent with the rest of the build system like gyp files, and it's functionally equivalent):

```diff
diff --git a/Makefile.am b/Makefile.am
index 0a2b246..2d2bdb1 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -36,7 +36,7 @@ libuv_la_SOURCES = src/fs-poll.c \
 if SUNOS
 # Can't be turned into a CC_CHECK_CFLAGS in configure.ac, it makes compilers
 # on other platforms complain that the argument is unused during compilation.
-libuv_la_CFLAGS += -pthread
+libuv_la_CFLAGS += -pthreads
 endif
 
 if WINNT
@@ -134,6 +134,13 @@ test_run_tests_CFLAGS =
 else
 test_run_tests_CFLAGS = -Wno-long-long
 endif
+
+if SUNOS
+# Can't be turned into a CC_CHECK_CFLAGS in configure.ac, it makes compilers
+# on other platforms complain that the argument is unused during compilation.
+test_run_tests_CFLAGS += -pthreads
+endif
+
 test_run_tests_LDFLAGS =
 test_run_tests_SOURCES = test/blackhole-server.c \
                          test/dns-server.c \
```

This is the diff of the preprocessor output for `test-fs.c` between not passing `-pthreads` and passing `-pthreads` on SmartOS:

```diff
--- test-fs-autotools-no-pthread.i      2016-09-14 19:06:57.529770010 +0000
+++ test-fs-autotools-with-pthreads.i   2016-09-14 19:07:30.193570124 +0000
@@ -20,8 +20,8 @@
 # 43 "/usr/include/sys/errno.h" 3 4
 #pragma ident "%Z%%M%  %I%     %E% SMI"
 # 42 "/usr/include/errno.h" 2 3 4
-# 58 "/usr/include/errno.h" 3 4
-extern int errno;
+# 55 "/usr/include/errno.h" 3 4
+extern int *___errno();
 # 26 "./include/uv-errno.h" 2
 # 49 "./include/uv.h" 2
 # 1 "./include/uv-version.h" 1
@@ -891,9 +891,7 @@
 
 
 
-
-
-extern int getdate_err;
+extern int *_getdate_err_addr(void);
 # 246 "/usr/include/time.h" 3 4
         
 # 246 "/usr/include/time.h" 3 4
@@ -2433,6 +2431,9 @@
 void freeipsecalgent(struct ipsecalgent *ptr);
 # 366 "/usr/include/netdb.h" 3 4
 extern int h_errno;
+
+
+extern int *__h_errno(void);
 # 35 "./include/uv-unix.h" 2
 
 # 1 "/usr/include/termios.h" 1 3 4
@@ -7925,7 +7926,7 @@
       else
         result = read(fd, data, size < interval ? size : interval);
 
-    } while (result == -1 && errno == 4);
+    } while (result == -1 && (*(___errno())) == 4);
 
     kill(pid, 16);
     size -= result;
```

As you can see from the diff above, the result when passing `-pthreads` is that the reentrant `___errno` function is used instead of the non-reentrant `errno` variable when `errno` is used in the code.

**Comment by saghul** [comment]
@misterdjules Interesting! Thanks for getting to the bottom of this! That sounds good to me. Care to send a PR for review?

**Comment by misterdjules** [comment]
@saghul

> Care to send a PR for review?

Let me do that now :)

**Comment by misterdjules** [comment]
@saghul https://github.com/libuv/libuv/pull/1052.

**Comment by bwijen** [comment]
@misterdjules: Great job!
@saghul: Let's wait until #1052 lands, I will rebase my work on top of that (without the SmartOS workaround)

**Comment by bwijen** [comment]
@saghul: rebased on #1052

**Comment by saghul** [comment]
Thanks @bwijen!  I plan on landing this right after 1.10 goes out (hopefully this week).  This one was a long time in the making, so thanks a lot for following it all the way through. :heart:

**Comment by bwijen** [comment]
@saghul: rebased on v1.x, just for fun.

I think this year is a good year to land :)

**Comment by bwijen** [comment]
Hi @saghul, just did another rebase.

I would really like to see this land...

**Comment by bwijen** [comment]
Another rebase to resolve conflicts.
Any feedback on merge or not-to-merge is welcome!

**Comment by bwijen** [comment]
@saghul: Is there anything left to do?

Please respond.

**Comment by bwijen** [comment]
Rebased to solve conficts...

**Comment by bwijen** [comment]
@bnoordhuis : Care to take a look?

**Comment by cjihrig** [comment]
This needs to be rebased again.

**Comment by bwijen** [comment]
@cjihrig: rebased to resolve conflicts

I would love to see this land in the near future...

**Comment by vtjnash** [comment]
Agreed. :)

File: test/test-fs.c [hunk 18]
```diff
@@ -2947,3 +2947,119 @@ TEST_IMPL(fs_null_req) {
...
+  int result;
+
+  iovs = malloc(sizeof(*iovs) * iovcount);
+  ASSERT(iovs != NULL);
+
+  ctx.doread = doread;
+  ctx.interval = 1000;
+  ctx.size = sizeof(test_buf) * iovcount;
+  ctx.data = malloc(ctx.size);
+  buffer = malloc(ctx.size);
```
Comment (santigimeno) [inline test/test-fs.c]: check these mallocs succeed

File: Makefile.am (hunk 5 above)
Comment (santigimeno) [reply Makefile.am]: This needs to be addressed

File: test/test-fs.c [hunk 19]
```diff
@@ -2947,3 +2947,119 @@ TEST_IMPL(fs_null_req) {
...
+struct thread_ctx {
+  int fd;
+  char *data;
+  int size;
+  int interval;
+  int doread;
+};
+
+static void thread_main(void* arg) {
+  struct thread_ctx *ctx = (struct thread_ctx*)arg;
```
Comment (santigimeno) [inline test/test-fs.c]: style: separate declaration from assignment

**Comment by santigimeno** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/479/

@saghul does this still LGTY?
@bnoordhuis could you take a look?

Thanks

File: Makefile.am (hunk 5 above)
Comment (bwijen) [reply Makefile.am]: @santigimeno: This commit already removes it from uv.gyp, what else needs to be addressed?

Comment (santigimeno) [reply Makefile.am]: @bwijen disregard that comment, I missed the change, sorry.

**Comment by santigimeno** [comment]
@libuv/collaborators could anyone take a look at this? Thanks!

**Comment by vtjnash** [comment]
@santigimeno This seems to be stalled again. Any chance you could land it while incorporating the style changes you marked above?

**Comment by vtjnash** [comment]
Monthly bump, hoping this bugfix can finally land.

**Comment by santigimeno** [comment]
Okay. If nobody objects, I'll land this on tuesday.

**Comment by santigimeno** [comment]
Added style fixups here: https://github.com/santigimeno/libuv/tree/santi_pr640.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/646/

**Comment by santigimeno** [comment]
One last CI after rebasing to current `v1.x`: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/649/

**Comment by santigimeno** [comment]
Landed in https://github.com/libuv/libuv/commit/14bfc27e641aff178c431083c0c0eada4d6f02dd. Thanks and sorry for the delay.

**Comment by bnoordhuis** [comment]
NB: we had to revert this commit, see #1717 and #1720.

**Pull request closed without merging**